File: verifier/src/jcs.test.ts
// SPDX-License-Identifier: Apache-2.0
import { describe, it, expect } from "vitest";
import { canonicalize } from "./jcs";

describe("canonicalize (RFC 8785 subset)", () => {
  it("serializes primitives", () => {
    expect(canonicalize(null)).toBe("null");
    expect(canonicalize(true)).toBe("true");
    expect(canonicalize(false)).toBe("false");
    expect(canonicalize(0)).toBe("0");
    expect(canonicalize(42)).toBe("42");
    expect(canonicalize(-7)).toBe("-7");
    expect(canonicalize("hi")).toBe('"hi"');
    expect(canonicalize("")).toBe('""');
  });

  it("sorts object keys by UTF-16 code-unit order", () => {
    expect(canonicalize({ b: 1, a: 2, c: 3 })).toBe('{"a":2,"b":1,"c":3}');
    // Capitals sort before lowercase (A=0x41 < a=0x61).
    expect(canonicalize({ a: 1, A: 2 })).toBe('{"A":2,"a":1}');
    // Digit keys sort before letters (0x30 < 0x41).
    expect(canonicalize({ z: 1, "1": 2 })).toBe('{"1":2,"z":1}');
  });

  it("preserves array order and never sorts it", () => {
    expect(canonicalize([3, 1, 2])).toBe("[3,1,2]");
    expect(canonicalize(["b", "a"])).toBe('["b","a"]');
    expect(canonicalize([])).toBe("[]");
  });

  it("nests objects and arrays with sorted keys at every level", () => {
    const v = { outer: { z: [1, { y: 1, x: 2 }], a: "v" }, first: true };
    expect(canonicalize(v)).toBe(
      '{"first":true,"outer":{"a":"v","z":[1,{"x":2,"y":1}]}}',
    );
  });

  it("escapes strings with standard JSON escaping, no normalization", () => {
    expect(canonicalize('a"b')).toBe('"a\\"b"');
    expect(canonicalize("a\\b")).toBe('"a\\\\b"');
    expect(canonicalize("line\nfeed")).toBe('"line\\nfeed"');
    expect(canonicalize("tab\tend")).toBe('"tab\\tend"');
    // Control character below U+0020 escapes as \u00XX.
    expect(canonicalize("")).toBe('"\\u0001"');
    // Non-ASCII stays literal (no escaping, no NFC/NFD normalization).
    const s = "é"; // e + combining acute, NOT precomposed
    expect(canonicalize(s)).toBe('"' + s + '"');
    expect(canonicalize("é")).toBe('"é"'); // precomposed stays distinct
  });

  it("rejects non-finite numbers", () => {
    expect(() => canonicalize(Infinity)).toThrow();
    expect(() => canonicalize(-Infinity)).toThrow();
    expect(() => canonicalize(NaN)).toThrow();
  });

  it("rejects undefined values", () => {
    expect(() => canonicalize(undefined)).toThrow();
    expect(() => canonicalize({ a: undefined })).toThrow();
  });

  it("matches a known small credential-shaped fixture", () => {
    const cred = {
      "@context": "https://c2pa.org/ns/manifest/1.4",
      type: "ContentCredential",
      asset: { sha256: "ab", size: 3, mime: "text/markdown" },
      signature: { alg: "Ed25519", public_key: "k" },
    };
    expect(canonicalize(cred)).toBe(
      '{"@context":"https://c2pa.org/ns/manifest/1.4",' +
        '"asset":{"mime":"text/markdown","sha256":"ab","size":3},' +
        '"signature":{"alg":"Ed25519","public_key":"k"},' +
        '"type":"ContentCredential"}',
    );
  });
});

File: c2pa/jcs_test.go
// SPDX-License-Identifier: Apache-2.0

package c2pa

import "testing"

// These vectors mirror verifier/src/jcs.test.ts one for one. The two
// canonicalizers have to agree on every byte, so they are tested
// against the same expectations in both languages.
func TestCanonicalJSONVectors(t *testing.T) {
	cases := []struct{ in, want string }{
		{`null`, `null`},
		{`true`, `true`},
		{`false`, `false`},
		{`0`, `0`},
		{`42`, `42`},
		{`-7`, `-7`},
		{`"hi"`, `"hi"`},
		{`""`, `""`},

		// Object keys sort by UTF-16 code-unit order.
		{`{"b":1,"a":2,"c":3}`, `{"a":2,"b":1,"c":3}`},
		{`{"a":1,"A":2}`, `{"A":2,"a":1}`},
		{`{"z":1,"1":2}`, `{"1":2,"z":1}`},

		// Array order is preserved and never sorted.
		{`[3,1,2]`, `[3,1,2]`},
		{`["b","a"]`, `["b","a"]`},
		{`[]`, `[]`},

		// Nesting sorts keys at every level.
		{
			`{"outer":{"z":[1,{"y":1,"x":2}],"a":"v"},"first":true}`,
			`{"first":true,"outer":{"a":"v","z":[1,{"x":2,"y":1}]}}`,
		},

		// String escaping: the mandatory short forms, control
		// characters as \u00xx, non-ASCII literal and unnormalized.
		{`"a\"b"`, `"a\"b"`},
		{`"a\\b"`, `"a\\b"`},
		{"\"line\\nfeed\"", `"line\nfeed"`},
		{"\"tab\\tend\"", `"tab\tend"`},
		{`"\u0001"`, `"\u0001"`},
		{`"é"`, "\"é\""},
		{`"é"`, "\"é\""},
		// Characters outside the basic multilingual plane stay literal.
		{`"🚀"`, "\"\U0001F680\""},

		// A credential-shaped fixture, matching the TypeScript vector.
		{
			`{"@context":"https://c2pa.org/ns/manifest/1.4","type":"ContentCredential",` +
				`"asset":{"sha256":"ab","size":3,"mime":"text/markdown"},` +
				`"signature":{"alg":"Ed25519","public_key":"k"}}`,
			`{"@context":"https://c2pa.org/ns/manifest/1.4",` +
				`"asset":{"mime":"text/markdown","sha256":"ab","size":3},` +
				`"signature":{"alg":"Ed25519","public_key":"k"},` +
				`"type":"ContentCredential"}`,
		},
	}
	for _, c := range cases {
		got, err := canonicalJSON([]byte(c.in))
		if err != nil {
			t.Fatalf("canonicalJSON(%s): %v", c.in, err)
		}
		if got != c.want {
			t.Fatalf("canonicalJSON(%s) = %s, want %s", c.in, got, c.want)
		}
	}
}

func TestCanonicalJSONRejectsUnsafeNumbers(t *testing.T) {
	for _, in := range []string{`9007199254740993`, `1.5`, `-2.25`} {
		if got, err := canonicalJSON([]byte(in)); err == nil {
			t.Fatalf("canonicalJSON(%s) = %s, want an error", in, got)
		}
	}
}

func TestCanonicalJSONRejectsMalformedInput(t *testing.T) {
	for _, in := range []string{``, `{`, `{"a":}`} {
		if _, err := canonicalJSON([]byte(in)); err == nil {
			t.Fatalf("canonicalJSON(%q) accepted malformed input", in)
		}
	}
}
